src/descriptor/create_descriptor.rs

```diff
@@ -0,0 +1,468 @@
+/// Miniscript should not use other pushes apart from PUSHNUM_1. This will err on reciving anything
+/// which is not PUSHBYTES, OR PUSHNUM_1 as other things are expected to happen in Miniscript
+///
+/// NOTE: Miniscript pushes should only be either boolean, 1 or 0, signatures, and hash preimages.
+/// As per the current implementation, PUSH_NUM2 results in an error
+fn instr_to_stackelem<'txin>(ins: &Instruction<'txin>) -> Result< StackElement<'txin>, Error> {
+    match *ins {
+        //Also covers the dissatisfied case as PushBytes0
+        Instruction::PushBytes(v) => Ok(StackElement::from(v)),
+        Instruction::Op(opcodes::all::OP_PUSHNUM_1) => Ok(StackElement::Satisfied),
+        _ => Err(Error::BadDescriptor),
+    }
+}
+
+/// Helper function which splits the scriptsig into 2 parts returns the corresponding elements.
+/// Usually used for scripts which have top element as Pk (p2pkh) or redeem script(p2sh).
+/// Converts the other script elements into Vec<StackElement>
+fn parse_scriptsig_top(script_sig: &bitcoin::Script) -> Result<(Vec<u8>, Vec<StackElement>), Error>
+{
+    let instructions : Vec<Instruction> = script_sig.iter(true).collect();
+    if let Some((pk_bytes, witness)) = instructions.split_last(){
+        let stack: Result< Vec<StackElement>, Error> = witness
+            .iter()
+            .map(|instr| instr_to_stackelem(instr))
+            .collect();
+
+        if let &Instruction::PushBytes(ref top) = pk_bytes {
+            Ok((top.to_vec(), stack?))
+        }
+        else {
+            Err(Error::BadDescriptor)
+        }
+    } else{
+        Err(Error::BadDescriptor)
+    }
+}
+
+
+/// Creates a pk descriptor based on scriptsig and script_pubkey.
+/// Pushes all remaining witness elements into witness<StackElement>
+fn verify_p2pk<'txin>(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &'txin bitcoin::Script,
+    witness: & [Vec<u8>],
+) -> Result<(Descriptor<PublicKey, hash160::Hash>, Vec<StackElement<'txin>>), Error>
+{
+    let script_pubkey_len = script_pubkey.len();
+    let pk_bytes = &script_pubkey.to_bytes();
+    let pk = bitcoin::PublicKey::from_slice(&pk_bytes[1..script_pubkey_len-1]).unwrap();
+
+    let stack: Result< Vec<StackElement>, Error> = script_sig.iter(true)
+        .map(|instr| instr_to_stackelem(&instr))
+        .collect();
+    if !witness.is_empty(){
+        Err(Error::BadDescriptor)
+    } else {
+        Ok((Descriptor::Pk(pk), stack?))
+    }
+}
+/// Helper to create a wpkh descriptor based on script_pubkey and witness. Validates the pubkey hash
+/// and pushes rest of witness to Vec<StackElement> as is.
+/// This does not check the signature, only creates the corresponding descriptor
+fn verify_p2wpkh<'txin>(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &bitcoin::Script,
+    witness: &'txin [Vec<u8>],
+) -> Result<(PublicKey, Vec<StackElement<'txin>>), Error>
+{
+    //script_sig must be empty
+    if *script_sig != Script::new(){
```

Comment:
It's a bit more idiomatic to use `!script_sig.is_empty()` rather than comparing against `Script::new`.